### PR TITLE
fix(ast-spec): use `Expression` in argument of `ThrowStatement`

### LINKED ISSUE #9610: Bug: ThrowStatement's argument should be Expression
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

ast-spec

### Issue

According to the ECMAScript specification, ThrowStatement is required to have an expression as argument:

https://tc39.es/ecma262/#sec-throw-statement

But in ast-spec, the argument is exposed as a statement or null:

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/ast-spec/src/statement/ThrowStatement/spec.ts

```
import type { AST_NODE_TYPES } from '../../ast-node-types';
import type { BaseNode } from '../../base/BaseNode';
import type { TSAsExpression } from '../../expression/TSAsExpression/spec';
import type { Statement } from '../../unions/Statement';

export interface ThrowStatement extends BaseNode {
  type: AST_NODE_TYPES.ThrowStatement;
  argument: Statement | TSAsExpression | null;
}
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/types` | `7.16.1` |

### PR DESCRIPTION
**Pull request by dak2**
## PR Checklist

- [x] Addresses an existing open issue: fixes #9610
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

- Use `Expression` in argument of `ThrowStatement`
- Remove statement and null type from ThrowStatement

### CODE REVIEW

**Review by bradzacher — CHANGES_REQUESTED** [review]
This isn't correct! The type of `argument` should just be `Expression` as per the title of the issue.

**Comment by dak2** [comment]
@bradzacher 

Thanks! Looks like I was mistaken, I fixed it to use Expression.

**Comment by dak2** [comment]
I don't know why CI failed. Perhaps there is an error when running `yarn`, but I can't reproduce it locally.

**Comment by Josh-Cena** [comment]
Here's the relevant log in CI:

```
  ➤ YN0000: │ @typescript-eslint/typescript-eslint@workspace:. STDOUT > nx run eslint-plugin:build
  Error: ➤ YN0000: │ @typescript-eslint/typescript-eslint@workspace:. STDOUT src/rules/no-throw-literal.ts(93,28): error TS2345: Argument of type 'TSAsExpression | Statement | null' is not assignable to parameter of type 'Node'.
  ➤ YN0000: │ @typescript-eslint/typescript-eslint@workspace:. STDOUT   Type 'null' is not assignable to type 'Node'.
  Error: ➤ YN0000: │ @typescript-eslint/typescript-eslint@workspace:. STDOUT src/rules/only-throw-error.ts(92,28): error TS2345: Argument of type 'TSAsExpression | Statement | null' is not assignable to parameter of type 'Node'.
```

**Review by bradzacher — APPROVED** [review]
Brilliant - thanks for this!
Once we fix the build errors we're good to land this

**Comment by bradzacher** [comment]
cc @JamesHenry - I think that NX is improperly caching the build here.
The AST spec has changed so that the property is just `argument: Expression`, yet the compilation step from `ast-spec` src to `types/src/generated/ast-spec.ts` is reusing the old output.

**Comment by dak2** [comment]
@bradzacher @Josh-Cena 
Thanks for your help! 
I think rebuild without cache before run `yarn install --immutable --inline-builds` in GitHub Actions. 
Any concerns about this approach? Or is there another effective way?

**Comment by Josh-Cena** [comment]
@dak2 You don't need to worry about CI failures. As long as your code is valid, the burden of fixing the CI is on the team.

File: .github/actions/prepare-install/action.yml [hunk 1]
```diff
@@ -55,9 +55,16 @@ runs:
         restore-keys: |
           yarn-download-cache-
 
+    - name: Check if ast-spec files changed
+      id: changed-files
+      uses: tj-actions/changed-files@v41
+      with:
+        files: packages/ast-spec/src/**
```
Comment (Josh-Cena) [inline .github/actions/prepare-install/action.yml:62]: Can you revert this or move this to a separate PR/issue if it's necessary?

Comment (dak2) [reply .github/actions/prepare-install/action.yml]: @Josh-Cena 

>  You don't need to worry about CI failures. As long as your code is valid, the burden of fixing the CI is on the team.

Thanks!!

> Can you revert this or move this to a separate PR/issue if it's necessary?

Undo file changes in github actions.

https://github.com/typescript-eslint/typescript-eslint/pull/9632/commits/a8995448520ef510b896adbd21e3c8a6c96eefcc

**Comment by JamesHenry** [comment]
@bradzacher Sorry I missed your comment here, let's double check and make sure we have the cache inputs set up correctly for this case

**Review by bradzacher — APPROVED** [review]

**Comment by JamesHenry** [comment]
The reason for the failure was that we were telling nx to exclude test files from build cache inputs. `spec.ts` is a super common naming convention for test files and so that's why it was included in the pattern, but our more unique "ast spec" convention was therefore caught in the crossfire.

In our repo we are actually consistent with our usage of `test.ts` for test files, so the solution here is to stop classing `spec.ts` as test files in the build inputs (and I did some minor consistency cleanup for old references to spec in the test context)

**Comment by JamesHenry** [comment]
Thanks again @dak2!
